File: KotlinStudy/step05httprequest/src/main/java/com/example/step05httprequest/MainActivity.java
package com.example.step05httprequest;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.example.step05httprequest.databinding.ActivityMainBinding;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;


public class MainActivity extends AppCompatActivity {

    ActivityMainBinding binding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        setContentView(R.layout.activity_main);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());


        binding.requestBtn.setOnClickListener(view -> {
            String url = "https://acornacademy.co.kr";
            makeHttpRequest(url);
        });

        binding.requestBtn2.setOnClickListener(view -> {
            String url = "https://acornacademy.co.kr";
            makeHttpRequest2(url);
        });
    }

    // 다른 스레드에서 메세지를 받아서 ui를 업데이트할 handler 객체
    Handler handler = new Handler(Looper.getMainLooper()){

        @Override
        public void handleMessage(@NonNull Message msg) {
            // main thread
            String data = (String)msg.obj;
            binding.editText.setText(data);
        }
    };

    /*
        request 1은 requestThread 를 이용하는 방법
        request 2는 runOnUiThread 를 이용하는 방법
     */


    // http 요청을 하는 메소드
    public void makeHttpRequest(String url) {
        new RequestThread(url).start();         //새로운 스레드에서 http 요청을 한다. - main thread에서 하면 안됨
    }

    // Executor 객체를 이용해서 새로운 스레드에서 작업하기
    public void makeHttpRequest2(String urlAddr){
        Executor executor = Executors.newSingleThreadExecutor();
        executor.execute(()->{
            //여기는 새로운 스레드 이다
            //문자열을 누적할 객체
            StringBuilder builder = new StringBuilder();

            try {
                //요청 url 을 생성자의 인자로 전달하면서 URL 객체를 생성한다
                URL url = new URL(urlAddr);
                //URLConnection 객체를 원래 type (자식 type) 으로 casting 해서 받는다.
                HttpURLConnection conn = (HttpURLConnection)url.openConnection();
                if(conn != null){
                    conn.setConnectTimeout(20000); //응답을 기다리는 최대 대기 시간
                    conn.setRequestMethod("GET");// 요청 메소드 설정 (Default 는 GET)
                    conn.setUseCaches(false);//케쉬 사용 여부
                    //응답 코드를 읽어온다.
                    int responseCode=conn.getResponseCode();
                    if(responseCode == HttpURLConnection.HTTP_OK) { //정상 응답이라면 (200)
                        //문자열을 읽어들일수 있는 객체의 참조값 얻어오기
                        BufferedReader br =
                                new BufferedReader(new InputStreamReader(conn.getInputStream()));
                        //반복문 돌면서
                        while(true){
                            //문자열을 한줄씩 읽어 들인다.
                            String line = br.readLine();
                            if(line == null)break;
                            //StringBuilder 객체에 누적 시키기
                            builder.append(line);
                        }
                    }
                }
            }catch(Exception e){
                Log.e("MainActivity", e.getMessage());
            }//try~catch

            // UI 스레드에서 동작하는 함수에서 UI 업데이트 하기
            runOnUiThread(()->{
                binding.editText.setText(builder.toString());
            });
        });

    }

    class RequestThread extends Thread {     // 스레드 클래스
        String url;

        public RequestThread(String url) { // constructor
            this.url = url;
        }

        @Override
        public void run() {
            //문자열을 누적할 객체
            StringBuilder builder = new StringBuilder();

            try {
                //요청 url 을 생성자의 인자로 전달하면서 URL 객체를 생성한다
                URL url = new URL(this.url);
                //URLConnection 객체를 원래 type (자식 type) 으로 casting 해서 받는다.
                HttpURLConnection conn = (HttpURLConnection)url.openConnection();
                if(conn != null){
                    conn.setConnectTimeout(20000); //응답을 기다리는 최대 대기 시간
                    conn.setRequestMethod("GET");// 요청 메소드 설정 (Default 는 GET)
                    conn.setUseCaches(false);//케쉬 사용 여부
                    //응답 코드를 읽어온다.
                    int responseCode=conn.getResponseCode();
                    if(responseCode == HttpURLConnection.HTTP_OK) { //정상 응답이라면 (200)
                        //문자열을 읽어들일수 있는 객체의 참조값 얻어오기
                        BufferedReader br =
                                new BufferedReader(new InputStreamReader(conn.getInputStream()));
                        //반복문 돌면서
                        while(true){
                            //문자열을 한줄씩 읽어 들인다.
                            String line = br.readLine();
                            if(line == null)break;
                            //StringBuilder 객체에 누적 시키기
                            builder.append(line);
                        }
                    }
                }
            }catch(Exception e){
                Log.e("MainActivity", e.getMessage());
            }//try~catch

            String result = builder.toString();
            Log.d("response data", result);

            Message msg = new Message();
            msg.what = 0;

            msg.obj = builder.toString();
            handler.sendMessage(msg);
        }

    }
}

File: KotlinStudy/step06coroutine/src/main/java/com/example/step06coroutine/MainActivity.kt
package com.example.step06coroutine

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

class MainActivity : AppCompatActivity() {

    var index = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val startBtn:Button = findViewById(R.id.startBtn)
        val resultText:TextView = findViewById(R.id.result)

        val getBtn:Button = findViewById(R.id.getBtn)
        val resultText2:TextView = findViewById(R.id.result2)

        startBtn.setOnClickListener{
            /*
                lifecycleScope.launch - activity의 생명주기 내에서만 동작
                                        activity가 비활성화되면 자동으로 취소됨

                Dispatchers.Main      - Main(UI) thread 내에서만 동작하도록 한다
                                        결과 값을 이용해서 UI를 수정하고자 할 때 사용한다
             */
            lifecycleScope.launch(Dispatchers.Main) {

                val result = doSomething() + index++
                resultText.setText(result)
            }
        }

        getBtn.setOnClickListener{
            lifecycleScope.launch(Dispatchers.Main) {
                val result = getMembers()
                resultText2.setText(result)
                // resultText2.text = result
            }
        }
    }

    suspend fun getMembers():String {
        val result = withContext(Dispatchers.IO){

            val builder = StringBuilder()

            try {
                //요청 url 을 생성자의 인자로 전달하면서 URL 객체를 생성한다
                val url = URL("http://192.168.0.205:8888/api/members")
                //URLConnection 객체를 원래 type (자식 type) 으로 casting 해서 받는다.
                val conn = url.openConnection() as HttpURLConnection
                if (conn != null) {
                    conn.connectTimeout = 20000 //응답을 기다리는 최대 대기 시간
                    conn.requestMethod = "GET" // 요청 메소드 설정 (Default 는 GET)
                    conn.useCaches = false //케쉬 사용 여부
                    //응답 코드를 읽어온다.
                    val responseCode = conn.responseCode
                    if (responseCode == HttpURLConnection.HTTP_OK) { //정상 응답이라면 (200)
                        //문자열을 읽어들일수 있는 객체의 참조값 얻어오기
                        val br =
                            BufferedReader(InputStreamReader(conn.inputStream))
                        //반복문 돌면서
                        while (true) {
                            //문자열을 한줄씩 읽어 들인다.
                            val line = br.readLine() ?: break
                            //StringBuilder 객체에 누적 시키기
                            builder.append(line)
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("MainActivity", e.message!!)
            } //try~catch

            builder.toString()
        }
        return result
    }

    // 오래걸리는 작업 메소드
    suspend fun doSomething():String {
        /*
            withContext     - Context는 activity임 -> activity의 생명 주기 안에서 동작
            Dispatchers.ID  - Input Output 에 적합한 thread 상에서 동작하도록 함
                            - http request, file io 작업에 적합
         */
        val result = withContext(Dispatchers.IO) {
            Thread.sleep(3000)
            "[result str]"
        }
        return result
    }
}